### PR TITLE
Add weather via city

### PR DESCRIPTION
**Pull request by preckrasno**
Added option to fetch weather via city name and iso 3166 country code

### CODE REVIEW

File: DatWeatherDoe/API/Repository/City/CityValidator.swift [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    
+    init(city: String) {
+        self.city = city
+    }
+        
+    func validate() throws {
+        let isCityPresent = !city.isEmpty
+        let isCityPresentWithCountryCode = city.split(separator: ",").count == 2
+        let isValid = isCityPresent && isCityPresentWithCountryCode
+        if(!isValid) {
```
[the remark this answers is not in the document]
Comment (preckrasno) [reply DatWeatherDoe/API/Repository/City/CityValidator.swift]: fixed

File: DatWeatherDoe/API/Repository/City/CityValidator.swift [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+            throw WeatherError.zipCodeIncorrect
```
[the remark this answers is not in the document]
Comment (preckrasno) [reply DatWeatherDoe/API/Repository/City/CityValidator.swift]: fixed

File: DatWeatherDoeTests/API/Repository/City/CityWeatherURLBuilderTests.swift [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+//  Created by preckrasno on 14.02.2023.
+//  Copyright © 2023 Inder Dhir. All rights reserved.
+//
+
+import XCTest
+@testable import DatWeatherDoe
+
+class CityWeatherURLBuilderTests: XCTestCase {
+
+    override func setUpWithError() throws {
```
[the remark this answers is not in the document]
Comment (preckrasno) [reply DatWeatherDoeTests/API/Repository/City/CityWeatherURLBuilderTests.swift]: fixed

**Comment by preckrasno** [comment]
> Left some comments.
> 
> Please note that 'zipcode' and 'city' options may be removed for the next major release (4.0.0) but we can bring it in for 3.x.x releases.
> 
> Ideally, the app needs to cover 2 use cases:
> 
> * Automatic location fetching that updates as the user changes locations, and
> * Manual location based on something like city or lat/long.
> 
> System location and lat/long input cover both of these cases. Separately, there has been an issue with city Id coming back from the API for the zip code case where "See Full Weather" always shows the weather for the wrong location. This issue is not present for the "city" case though.

Regarding the goal of this PR. I had issues with the default location-based and lat/long options, as they showed weather for the wrong locality. Additionally, I found that openweathermap was not handling my zip code and country combination properly. This led me to believe that adding an option to retrieve weather data by city would be beneficial for the app.

Thank you for your comments and feedback, and I look forward to any further suggestions or changes that may be needed to improve this feature.

**Review by inderdhir — APPROVED** [review]

**Pull request merged**